### PR TITLE
Makefile: clean git branch before building

### PR DESCRIPTION
**Pull request by runcom**
Fix: https://github.com/docker/docker/issues/18268

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

**Comment by runcom** [comment]
ping @aaronlehmann not sure which other chars are invalid as a tag - using the tag regexp in distribution/ code didn't work as the *invalid* tag is somehow valid (I guess the validation is done as a whole and so the tag with "/" is valid but when taking into account a full ref name it's invalid)

**Comment by aaronlehmann** [comment]
@runcom: Which regexp were you using? This one should be the definitive grammar for tags: `https://github.com/docker/distribution/blob/master/reference/regexp.go#L41`

(expressed as a regexp: `^[\w][\w.-]{0,127}$`)

**Comment by runcom** [comment]
Right, if you run "pr/upstream/number"  against that regexp it's not returning false :/

**Comment by runcom** [comment]
Wait, let me double check again but I remember it wasn't returning false.

**Comment by runcom** [comment]
```
[amurdaca@localhost docker]$ [[ "prupstream14324" =~ ^[\w][\w.-]{0,127}$ ]] && echo "match"
[amurdaca@localhost docker]$ [[ "pr/upstream/14324" =~ ^[\w][\w.-]{0,127}$ ]] && echo "match"
[amurdaca@localhost docker]$
```

**Comment by aaronlehmann** [comment]
It looks like bash doesn't like `\w` inside `[]`. Maybe that's a perl extension that got picked up by Go...

I got the following regexp to work:

```
$ [[ "prupstream/14324" =~ ^[[:alnum:]_][[:alnum:]_.-]{0,127}$ ]] && echo match
$ [[ "prupstream14324" =~ ^[[:alnum:]_][[:alnum:]_.-]{0,127}$ ]] && echo match
match
```

**Comment by runcom** [comment]
awesome, though now I'm not sure how to replace the wrong chars :confused:

File: Makefile [hunk 1]
```diff
@@ -40,8 +40,9 @@ DOCKER_MOUNT := $(if $(BIND_DIR),-v "$(CURDIR)/$(BIND_DIR):/go/src/github.com/do
 DOCKER_MOUNT := $(if $(DOCKER_MOUNT),$(DOCKER_MOUNT),-v "/go/src/github.com/docker/docker/bundles")
 
 GIT_BRANCH := $(shell git rev-parse --abbrev-ref HEAD 2>/dev/null)
-DOCKER_IMAGE := docker-dev$(if $(GIT_BRANCH),:$(GIT_BRANCH))
-DOCKER_DOCS_IMAGE := docker-docs$(if $(GIT_BRANCH),:$(GIT_BRANCH))
+GIT_BRANCH_CLEAN := $(shell echo $(GIT_BRANCH) | sed -e "s/\//-/g")
```
Comment (thaJeztah) [inline Makefile]: perhaps just

    s/[^[:alnum:]]/-/g

**Comment by thaJeztah** [comment]
LGTM, but leaving https://github.com/docker/docker/pull/22400#discussion_r61518626 as a suggestion

ping @aaronlehmann ptal if this LGTY

**Comment by aaronlehmann** [comment]
LGTM

Agree with @thaJeztah that it would be useful to cover more problematic characters than just `/`, though.

File: Makefile (hunk 1 above)
Comment (LK4D4) [inline Makefile]: I agree, @runcom could you change pls?

Comment (runcom) [inline Makefile]: Right

**Comment by runcom** [comment]
@LK4D4 @thaJeztah updated!

**Comment by vdemeester** [comment]
LGTM🐻

**Comment by thaJeztah** [comment]
thanks @runcom

**Pull request merged**